contracts/smart-account/modules/SessionValidationModules/ABISessionValidationModule.sol

```diff
@@ -0,0 +1,164 @@
+     * @param _userOpHash Hash of the User Operation to be validated.
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     * @param _sessionKeySignature Signature over the the _userOpHash.
+     * @return true if the _op is valid, false otherwise.
+     */
+    function validateSessionUserOp(
+        UserOperation calldata _op,
+        bytes32 _userOpHash,
+        bytes calldata _sessionKeyData,
+        bytes calldata _sessionKeySignature
+    ) external pure override returns (bool) {
+        require(
+            bytes4(_op.callData[0:4]) == EXECUTE_OPTIMIZED_SELECTOR ||
+                bytes4(_op.callData[0:4]) == EXECUTE_SELECTOR,
+            "ABISV Invalid Selector"
+        );
+
+        // we expect _op.callData to be `SmartAccount.execute(to, value, calldata)` calldata
+        (address destinationContract, uint256 callValue, ) = abi.decode(
+            _op.callData[4:], // skip selector
+            (address, uint256, bytes)
+        );
+
+        bytes calldata data;
+        {
+            //offset represents where does the inner bytes array start
+            uint256 offset = uint256(bytes32(_op.callData[4 + 64:4 + 96]));
+            uint256 length = uint256(
+                bytes32(_op.callData[4 + offset:4 + offset + 32])
+            );
+            data = _op.callData[4 + offset + 32:4 + offset + 32 + length];
+        }
+
+        return
+            validateSessionParams(
+                destinationContract,
+                callValue,
+                data,
+                _sessionKeyData,
+                new bytes(0)
+            ) ==
+            ECDSA.recover(
+                ECDSA.toEthSignedMessageHash(_userOpHash),
+                _sessionKeySignature
+            );
+    }
+
+    /**
+     * @dev validates that the call (destinationContract, callValue, funcCallData)
+     * complies with the Session Key permissions represented by sessionKeyData
+     * @param destinationContract address of the contract to be called
+     * @param callValue value to be sent with the call
+     * @param _funcCallData the data for the call. is parsed inside the SVM
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     * param _callSpecificData additional data, for example some proofs if the SVM utilizes merkle trees itself
+     * for example to store a list of allowed tokens or receivers
+     */
+    function validateSessionParams(
+        address destinationContract,
+        uint256 callValue,
+        bytes calldata _funcCallData,
+        bytes calldata _sessionKeyData,
+        bytes memory /*_callSpecificData*/
+    ) public pure virtual override returns (address) {
+        (address sessionKey, Permission memory permission) = abi.decode(
+            _sessionKeyData,
+            (address, Permission)
+        );
+
+        if (destinationContract != permission.destinationContract) {
+            revert("ABISV Wrong Destination");
```

Comment:
It is more correct to revert here. It provides richer information regarding what went wrong. also allows avoiding extra check of the return in the caller contract

Reply:
Returning and not reverting during validation phase, helps in simulating the transaction via EntryPoint simulation functions as it will not revert during simulation where we are just estimating gas values.